src/python/modules/ensembl/compara/plugins/TestDB.py

```diff
@@ -0,0 +1,249 @@
+        return sql_filter
+
+    def test_num_rows(self, variation: float = 0.0, group_by: Union[str, List] = None,
+                      filter_by: Union[str, List] = None) -> None:
+        """Compares the number of rows between reference and target tables.
+
+        If `group_by` is provided, the number of rows will be compared per group, applying the same variation
+        to all of them. If `filter_by` is provided, only the rows matching all the given conditions will be
+        compared.
+
+        Note:
+            The ceiling function is applied to round the allowed variation in order to compare two integers.
+            Thus, the test may pass even if the difference between two files is greater than the exact allowed
+            variation, e.g. the test will pass if the difference is 2 and the allowed variation is 1,4.
+
+        Args:
+            variation: Allowed variation between reference and target tables.
+            group_by: Group rows by column(s), and count the number of rows per group.
+            filter_by: Filter rows by one or more conditions (joined by the AND operator).
+
+        Raise:
+            TestDBException: If `group_by` is provided and the number of groups is different; or if the number
+                of rows differ for at least one group.
+
+        """
+        # Compose the sql query from the given parameters
+        sql_filter = self._get_sql_filter(filter_by)
+        if group_by:
+            if (group_by is None) or isinstance(group_by, list):
+                group_by = ", ".join(group_by)
+            # ORDER BY to ensure that the results are always in the same order (for the same groups)
+            sql_query = "SELECT {0}, COUNT(*) as nrows FROM {1} {2} GROUP BY {0} ORDER BY {0}".format(
+                group_by, self.table, sql_filter)
+        else:
+            sql_query = "SELECT COUNT(*) as nrows FROM {} {}".format(self.table, sql_filter)
+        # Get the number of rows for both databases
+        result = self.ref_db.execute(sql_query)
+        ref_data = pandas.DataFrame(result.fetchall(), columns=result.keys())
+        result = self.target_db.execute(sql_query)
+        target_data = pandas.DataFrame(result.fetchall(), columns=result.keys())
+        # Check if the size of the returned tables are the same
+        if ref_data.shape != target_data.shape:
+            expected = ref_data.shape[0]
+            found = target_data.shape[0]
+            # Note: the shape can only be different if group_by is given
+            message = "Different number of groups ({}) for table '{}'".format(group_by, self.table)
+            raise TestDBException(expected, found, sql_query, message)
+        # Check if the number of rows (per group) are the same
+        difference = abs(ref_data['nrows'] - target_data['nrows'])
+        allowed_variation = numpy.ceil(ref_data['nrows'] * variation)
+        failing_rows = difference > allowed_variation
```

Comment:
Then, indirectly, you would be applying the `floor` function instead. For very assorted groups, a small variability may allow no variability at all for the smallest groups, and I wasn't sure that is what we wanted. But it would be definitely faster (and less comment lines) without the `ceil` function.

Reply:
No. Say the reference number of rows is 50, and we allow a variation of 5%, i.e. 2.5
2 is ok, 3 is not. I guess that's still the ceiling, but the difference is the usage of `>` vs `>=`.
Your current implementation would allow 3, but it shouldn't.